### PR TITLE
Fix #849, allow CUSPARSE to use F16

### LINKED ISSUE #849: Support for CuSparseMatrixCSC{Float16} x CuVector{Float16}
**Describe the bug**

I get the "Performing scalar operations" warning when I try to do sparse matrix/dense vector multiplication with Float16 precision.
When I set `CUDA.allowscalar(false)`, this becomes "ERROR: scalar getindex is disallowed." and crashes.

When I try the same thing in Float32 precision, the code works fine.

Note: dense matrix/dense vector multiplication _does_ work in Float16. So this seems fixable. 

**To reproduce**

```julia

using CUDA
using SparseArrays
using LinearAlgebra

# This works:
A_d = CUDA.CUSPARSE.CuSparseMatrixCSC{Float32}(sprand(10000,10000))
v_d = CuArray{Float32}(randn(10000))

println(A_d * v_d)
# 10000-element CuArray{Float32, 1}:
# etc.

# This doesn't work:
A_d_16 = CUDA.CUSPARSE.CuSparseMatrixCSC{Float16}(sprand(10000,10000))
v_d_16 = CuArray{Float16}(randn(10000))

println(A_d_16 * v_d_16)
```

<details><summary>Manifest.toml</summary>
<p>

```
[[CUDA]]
version = "3.0.3"

[[GPUArrays]]
version = "6.2.2"

[[GPUCompiler]]
version = "0.11.4"

[[LLVM]]
version = "3.6.0"
```

</p>
</details>

**Expected behavior**

Multiplication should work in Float16, just as it does in Float32

**Version info**

Details on Julia: 

```
Julia Version 1.6.0
Commit f9720dc2eb (2021-03-24 12:55 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, sandybridge)
```

Details on CUDA:

```
CUDA toolkit 11.2.2, artifact installation
CUDA driver 11.3.0
NVIDIA driver 465.19.1

Libraries: 
- CUBLAS: 11.4.1
- CURAND: 10.2.3
- CUFFT: 10.4.1
- CUSOLVER: 11.1.0
- CUSPARSE: 11.4.1
- CUPTI: 14.0.0
- NVML: 11.0.0+465.19.1
- CUDNN: 8.10.0 (for CUDA 11.2.0)
- CUTENSOR: 1.2.2 (for CUDA 11.1.0)

Toolchain:
- Julia: 1.6.0
- LLVM: 11.0.1
- PTX ISA support: 3.2, 4.0, 4.1, 4.2, 4.3, 5.0, 6.0, 6.1, 6.3, 6.4, 6.5, 7.0
- Device support: sm_35, sm_37, sm_50, sm_52, sm_53, sm_60, sm_61, sm_62, sm_70, sm_72, sm_75, sm_80

1 device:
  0: NVIDIA GeForce RTX 2070 (sm_75, 7.537 GiB / 7.795 GiB available)

```

Thanks for looking at this!

### PR DESCRIPTION
**Pull request by kshyatt**
Needs tests but `cyclops` was broken.

### CODE REVIEW

File: lib/cublas/wrappers.jl [hunk 1]
```diff
@@ -133,6 +133,9 @@ for (jname, fname, elty) in ((:dot,:cublasDdot_v2,:Float64),
         end
     end
 end
+dot(n::Integer, x::DenseCuArray{Float16}, y::DenseCuArray{Float16}) = convert(Float16, dot(n, convert(DenseCuArray{Float32}, x), convert(DenseCuArray{Float32}, y)))
```
Comment (maleadt) [inline lib/cublas/wrappers.jl]: Maybe add a TODO here; `cublasDotEx` _does_ support F16.

File: lib/cublas/util.jl [hunk 2]
```diff
@@ -30,9 +30,9 @@ function bandex(A::AbstractMatrix,kl,ku)
     return B
 end
 
-const CublasFloat = Union{Float64,Float32,ComplexF64,ComplexF32}
-const CublasReal = Union{Float64,Float32}
-const CublasComplex = Union{ComplexF64,ComplexF32}
+const CublasFloat = Union{Float64,Float32,ComplexF64,ComplexF32,Float16,ComplexF16}
+const CublasReal = Union{Float16,Float64,Float32}
```
Comment (maleadt) [inline lib/cublas/util.jl]: Is this safe? Won't this lead to dispatching to CUBLAS wrappers, potentially wrongfully so when the API doesn't have F16 support and we should have used a generic Julia fallback?

File: lib/cusparse/conversions.jl [hunk 3]
```diff
@@ -152,20 +208,94 @@ for (cname,rname,elty) in ((:cusparseScsc2dense, :cusparseScsr2dense, :Float32),
         function CUDA.CuMatrix{$elty}(csr::CuSparseMatrixCSR{$elty}; ind::SparseChar='O')
             m,n = csr.dims
             denseA = CUDA.zeros($elty,m,n)
-            lda = max(1,stride(denseA,2))
-            cudesc = CuMatrixDescriptor(CUSPARSE_MATRIX_TYPE_GENERAL, CUSPARSE_FILL_MODE_LOWER, CUSPARSE_DIAG_TYPE_NON_UNIT, ind)
-            $rname(handle(), m, n, cudesc, nonzeros(csr),
-                   csr.rowPtr, csr.colVal, denseA, lda)
-            denseA
+            if version() >= v"11.1"
```
Comment (maleadt) [inline lib/cusparse/conversions.jl]: CUDA libraries are versioned independently starting with CUDA 11, and `version()` returns that CUSPARSE version. So the version check needs to be more specific probably, see https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#cuda-major-component-versions

File: lib/cusparse/generic.jl [hunk 4]
```diff
@@ -84,13 +84,15 @@ function mv!(transa::SparseChar, alpha::Number, A::Union{CuSparseMatrixBSR{T},Cu
     descA = CuSparseMatrixDescriptor(A)
     descX = CuDenseVectorDescriptor(X)
     descY = CuDenseVectorDescriptor(Y)
-
+    compute_type = T == Float16 && version() >= v"11.3" ? Float32 : T
```
Comment (maleadt) [inline lib/cusparse/generic.jl]: Same with these.

File: lib/cusparse/interfaces.jl [hunk 5]
```diff
@@ -29,7 +29,7 @@ for (taga, untaga) in tag_wrappers, (wrapa, transa, unwrapa) in op_wrappers
 
     @eval begin
         function LinearAlgebra.mul!(C::CuVector{T}, A::$TypeA, B::DenseCuVector{T},
-                                    alpha::Number, beta::Number) where {T <: BlasFloat}
+                                    alpha::Number, beta::Number) where {T <: Union{Float16, ComplexF16, BlasFloat}}
```
Comment (maleadt) [inline lib/cusparse/interfaces.jl:32]: Why no `CublasFloat` here?

File: lib/cublas/wrappers.jl (hunk 1 above)
Comment (kshyatt) [reply lib/cublas/wrappers.jl]: Oh cool, I can just use that instead on CUDA versions which support it, then.

File: lib/cusparse/conversions.jl (hunk 3 above)
Comment (kshyatt) [reply lib/cusparse/conversions.jl]: Oh shoot, thanks!

File: lib/cusparse/interfaces.jl (hunk 5 above)
Comment (kshyatt) [reply lib/cusparse/interfaces.jl:32]: I forgot 😅

File: lib/cublas/util.jl (hunk 2 above)
Comment (kshyatt) [reply lib/cublas/util.jl]: Yeah, I can remove this and add fallbacks. It's going to lead to a lot of code but I guess that's the CUDA lifestyle for you.

**Comment by kshyatt** [comment]
CI failed due to a weird end of instruction error.

**Comment by maleadt** [comment]
Yeah, https://github.com/JuliaLang/LinearAlgebra.jl/issues/815. I've set it to allow-failure on master, but let's just go ahead and merge this.

**Pull request merged**